# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.


3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [25]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv'
df = pd.read_csv(url)
df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [26]:
df_filtered = df[(df['Total Claim Amount'] > 1000) & (df['Response'] == 'Yes')]
df_filtered.head()


Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
189         189  OK31456  California             11009.130490      Yes   
236         236  YJ16163      Oregon             11009.130490      Yes   
419         419  GW43195      Oregon             25807.063000      Yes   
442         442  IP94270     Arizona             13736.132500      Yes   
587         587  FJ28407  California              5619.689084      Yes   

     Coverage             Education Effective To Date EmploymentStatus Gender  \
189   Premium              Bachelor           1/24/11         Employed      F   
236   Premium              Bachelor           1/24/11         Employed      F   
419  Extended               College           2/13/11         Employed      F   
442   Premium                Master           2/13/11         Disabled      F   
587   Premium  High School or Below           1/26/11       Unemployed      M   

     ...  Number of Open Complaints Number of Policies     Policy Type  \
189  ...                        0.0                  1  Corporate Auto   
236  ...                        0.0                  1    Special Auto   
419  ...                        1.0                  2   Personal Auto   
442  ...                        0.0                  8   Personal Auto   
587  ...                        0.0                  1   Personal Auto   

           Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
189  Corporate L3            Offer2          Agent         1358.400000   
236    Special L3            Offer2          Agent         1358.400000   
419   Personal L2            Offer1         Branch         1027.200000   
442   Personal L2            Offer1            Web         1261.319869   
587   Personal L1            Offer2            Web         1027.000029   

     Vehicle Class Vehicle Size Vehicle Type  
189     Luxury Car      Medsize          NaN  
236     Luxury Car      Medsize            A  
419     Luxury Car        Small            A  
442            SUV      Medsize            A  
587            SUV      Medsize            A  

[5 rows x 26 columns]

In [27]:
total_claim = round(df_filtered.groupby(['Policy Type', 'Gender']).sum()['Total Claim Amount'],2)
total_claim.head()

Policy Type     Gender
Corporate Auto  F          6830.40
                M          4684.60
Personal Auto   F         38875.32
                M         25032.95
Special Auto    F          1358.40
Name: Total Claim Amount, dtype: float64

In [ ]:
df_state = df[df['Policy Type'].notnull()]
df_state

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [32]:

state_counts = df_state['State'].value_counts()

df_state_filtered = df_state[df_state['State'].isin(state_counts[state_counts > 500].index)]

df_state_filtered 

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10904       10904  QI63521      Nevada             19228.463620       No   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   

       Coverage             Education Effective To Date EmploymentStatus  \
0         Basic               College           2/18/11         Employed   
1         Basic               College           1/18/11       Unemployed   
2         Basic              Bachelor           2/10/11         Employed   
3      Extended               College           1/11/11         Employed   
4       Premium              Bachelor           1/17/11    Medical Leave   
...         ...                   ...               ...              ...   
10904     Basic  High School or Below           2/24/11       Unemployed   
10905   Premium              Bachelor           1/19/11       Unemployed   
10906     Basic               College            1/6/11         Employed   
10907  Extended              Bachelor            2/6/11         Employed   
10908   Premium               College           2/13/11         Employed   

      Gender  ...  Number of Open Complaints Number of Policies  \
0          M  ...                        0.0                  9   
1          F  ...                        0.0                  1   
2          M  ...                        0.0                  2   
3          M  ...                        0.0                  2   
4          F  ...                        NaN                  7   
...      ...  ...                        ...                ...   
10904      M  ...                        0.0                  2   
10905      F  ...                        NaN                  7   
10906      F  ...                        0.0                  6   
10907      F  ...                        0.0                  2   
10908      F  ...                        4.0                  6   

          Policy Type        Policy  Renew Offer Type  Sales Channel  \
0      Corporate Auto  Corporate L3            Offer3          Agent   
1       Personal Auto   Personal L3            Offer4    Call Center   
2       Personal Auto   Personal L3            Offer3    Call Center   
3      Corporate Auto  Corporate L3            Offer2         Branch   
4       Personal Auto   Personal L2            Offer1         Branch   
...               ...           ...               ...            ...   
10904   Personal Auto   Personal L2            Offer1         Branch   
10905   Personal Auto   Personal L1            Offer3            Web   
10906   Personal Auto   Personal L3            Offer2         Branch   
10907  Corporate Auto  Corporate L3            Offer1            Web   
10908   Personal Auto   Personal L1            Offer1         Branch   

       Total Claim Amount  Vehicle Class Vehicle Size Vehicle Type  
0              292.800000  Four-Door Car      Medsize          NaN  
1              744.924331  Four-Door Car      Medsize          NaN  
2              480.000000            SUV      Medsize            A  
3              484.013411  Four-Door Car      Medsize            A  
4              707.925645  Four-Door Car      Medsize          NaN  
...                   ...            ...          ...          ...  
10904          897.600000   

In [33]:

lifetime_value_stats = df.groupby(['Education', 'Gender'])['Customer Lifetime Value'].agg(['max', 'min', 'median'])

lifetime_value_stats


max          min       median
Education            Gender                                       
Bachelor             F       73225.95652  1904.000852  5640.505303
                     M       67907.27050  1898.007675  5548.031892
College              F       61850.18803  1898.683686  5623.611187
                     M       61134.68307  1918.119700  6005.847375
Doctor               F       44856.11397  2395.570000  5332.462694
                     M       32677.34284  2267.604038  5577.669457
High School or Below F       55277.44589  2144.921535  6039.553187
                     M       83325.38119  1940.981221  6286.731006
Master               F       51016.06704  2417.777032  5729.855012
                     M       50568.25912  2272.307310  5579.099207

Conclusion:
Higher Education Levels: Generally, customers with a Bachelor's degree have the highest maximum CLV, with females showing the highest individual value (73,225.96).

Median CLV: Females tend to have slightly higher median CLV across most education levels, except for Master's degree holders, where males are slightly lower.

High School Impact: Surprisingly, customers with High School or Below education show high median CLV, especially males, with the highest median (6,286.73) and maximum CLV (83,325.38).

In [34]:

df['Effective To Date'] = pd.to_datetime(df['Effective To Date'])

df['Month'] = df['Effective To Date'].dt.strftime('%B')  

# Create a pivot table with states as rows and months as columns, counting the number of policies
policy_sales = df.pivot_table(index='State', columns='Month', values='Policy Type', aggfunc='count', fill_value=0)

policy_sales


/var/folders/9z/k5tvnzx17sb0wxf4lrs5qkwm0000gp/T/ipykernel_49618/3211365433.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['Effective To Date'] = pd.to_datetime(df['Effective To Date'])


Month       February  January
State                        
Arizona          929     1008
California      1634     1918
Nevada           442      551
Oregon          1344     1565
Washington       425      463

In [35]:


# Calculate the total number of policies sold by state
total_policies_by_state = policy_sales.sum(axis=1)

# Get the top 3 states with the highest number of policies sold
top_3_states = total_policies_by_state.nlargest(3).index

# Filter the original pivot table to include only the top 3 states
top_3_policy_sales = policy_sales.loc[top_3_states]

# Display the result
print(top_3_policy_sales)


Month       February  January
State                        
California      1634     1918
Oregon          1344     1565
Arizona          929     1008
